lib/message/encrypt.js

```diff
@@ -30,40 +57,34 @@ export default async function encryptMessage({
     }
 
     if (detached) {
-        // Add cleartext signature for backward compatibility
-        const { message } = sanitizedOptions;
-        if (message.fromStream) {
-            // TODO clone input stream -- sanitizedOptions.message.fromStream ? await readMessage({ clone(<input data>) }) : sanitizedOptions.message;
-            throw new Error('Encrypting and detached-signing a message created from a stream is not supported yet');
+        if (!armor && isStream(textData || binaryData)) {
+            throw new Error('Unsupported detached signature when streaming data and requesting non-armored result');
         }
-        const signature = await sign({
-            message,
+
+        const detachedSignatureBinary = await sign({
+            message: await createMessageFromPassiveStreamClone({ [dataType]: data }),
             signingKeys: sanitizedOptions.signingKeys,
             signingKeyIDs: sanitizedOptions.signingKeyIDs,
             signingUserIDs: sanitizedOptions.signingUserIDs,
             date: sanitizedOptions.date,
             config: sanitizedOptions.config,
-            format: 'object',
+            format: 'binary',
             detached: true
         });
-        result.signature = armor ? signature.armor() : signature;
-        if (message.fromStream) {
-            result.signature = await readToEnd(
-                transformPair(message.packets.write(), async (readable, writable) => {
-                    await Promise.all([pipe(result.signature, writable), readToEnd(readable).catch(() => {})]);
-                })
-            );
-        }
-        // Encrypt signature
+
         result.encryptedSignature = await encrypt({
             ...sanitizedOptions,
-            // TODO clone if streaming signature
-            message: await createMessage({ binary: signature.packets.write() })
+            message: await createMessage({ binary: cloneIfStream(detachedSignatureBinary) })
```

Comment:
And then here
```suggestion
            message: await createMessage({
                binary: isStream(detachedSignatureBinary) ? clone(detachedSignatureBinary) : detachedSignatureBinary
            })
```
Also, `clone` is very cheap for non-stream binary data (it just returns a subarray, it doesn't actually clone the data, since the goal of `stream.clone` is to be able to read twice) so it's also fine to unconditionally call clone. But OK, it's also fine to leave this ternary